cloudsync/smartsync.py

```diff
@@ -0,0 +1,186 @@
+                 shuffle: bool = False,
+                 prioritize: Callable[[int, str], int] = None):
+        self.requestset = set()
+        self.excludeset = set()
+        super().__init__(providers, storage, tag, shuffle, prioritize)
+
+    def smart_sync_ent(self, ent):
+        self.requestset.add(ent)
+        self.excludeset.discard(ent)
+
+    def smart_sync_path(self, remote_path) -> List[SyncEntry]:
+        # q: do we want to raise an exception if path isn't found?
+        # q: do we need to also sync path's parents all the way to the root?
+        #   a: probably, but it doesn't make sense to add them to the requestset, just sync them implicitly
+        #       or maybe it does make sense to add them to the requestset? that way they will get subscribed too
+        #       or maybe always sync all folders at all times, and only files are smartsynced
+        ents = self.lookup_path(REMOTE, remote_path)
+        for ent in ents:
+            self.requestset.add(ent)
+        return ents
+
+    def smart_unsync_ent(self, ent):
+        self.requestset.discard(ent)
+        self.excludeset.add(ent)
+
+    def smart_unsync_path(self, remote_path):
+        for ent in tuple(self.requestset):
+            if ent[REMOTE].path == remote_path:
+                self.state.smart_unsync_ent(ent)
+                break
+
+    def smart_listdir(self, local_path, remote_path):
+        prov = self.providers[REMOTE]
+        for ent, _ignored_rel_path in self.get_kids(remote_path, REMOTE):
+            remote_ent_path = ent[REMOTE].path
+            _ignored_remote_ent_parent, remote_ent_name = self.providers[REMOTE].split(ent[REMOTE].path)
+            if prov.paths_match(prov.dirname(remote_ent_path), remote_path):
+                ent_info = prov.info_oid(ent[REMOTE].oid)
+                retval = DirInfo(otype=ent[REMOTE].otype,
+                                 oid='',
+                                 hash=None,
+                                 path=self.providers[LOCAL].join(local_path, remote_ent_name),
+                                 size=ent_info.size,  # TODO: ent[REMOTE].size,
+                                 name=remote_ent_name,
+                                 mtime=0,  # TODO: ent[REMOTE].mtime,
+                                 shared=False,  # TODO: ent[REMOTE].shared,
+                                 readonly=False,  # TODO: ent[REMOTE].readonly
+                                 )
+                yield retval  # TODO: this needs to yield DirInfo, also translate back to local
+
+    @property
+    def changes(self):
+        # Filter the changes to just those that are in the requestset, or those whose get_latest is stale viz. change_time
+        changes = super().changes
+        for ent in self.requestset:
+            if ent[LOCAL].path in changes:
+                yield ent
+
+    @property
+    def _changeset(self):
+        folders = [ent for ent in self._changeset_storage if ent[REMOTE].otype == DIRECTORY or 
```

Comment:
The purpose of this is to include changed folders in the changeset, regardless of whether it has been requested. The default for folders is to sync, the default for files is to not sync

Reply:
I'm with that but I'm  confused on why we need `or ent[REMOTE].change > ent[REMOTE]._last_gotten` and if files can end up in the `folders` array

Reply:
I wrote that code a long time ago, so I'm definitely working to try to remember what was the logic was behind that. I suspect that the `folders` list is just poorly named, and it was intended that files should end up there. This allows for a `get_latest` on changed entries (folders and files alike) and the `changes` property limits syncing to items that are present in the `requestset`

Reply:
I think it makes sense to answer these questions by writing tests anyway... If it's confusing if this actually works in a particular scenario, I can write a test to assert the correct behavior in that scenario. You have a suggestion what that scenario is?

Reply:
I think my comment is more about structure than a particular sync scenario. I think you answered my second question that `folders` may just be poorly named. I still don't think I understand why this condition is in there though `ent[REMOTE].change > ent[REMOTE]._last_gotten`  and why we would do this in a smartsync context but not in a regular cloudsync context.

Reply:
This code has changed quite a bit, but the reason why we do this in smartsync and not regular cloudsync is because every changed row shows up in a the changeset during regular cloudsync, but in smartsync we want to automatically sync any changed folders, and we need to include rows matching `ent[REMOTE].change > ent[REMOTE]._last_gotten` so that there is a chance to `get_latest()` for those rows. The new version of this code is structured more simply, and has better commentary, I think.